src/server/json_family.cc

```diff
@@ -574,23 +574,49 @@ OpResult<vector<OptSizeT>> OpArrLen(const OpArgs& op_args, string_view key, Json
   return vec;
 }
 
-OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, string_view path) {
+OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, string_view path,
+                                   JsonPathV2 expression) {
   vector<OptBool> vec;
-  auto cb = [&vec](const auto&, JsonType& val) {
-    if (val.is_bool()) {
-      bool current_val = val.as_bool() ^ true;
-      val = current_val;
-      vec.emplace_back(current_val);
-    } else {
-      vec.emplace_back(nullopt);
+  if (std::holds_alternative<json::Path>(expression)) {
+    auto it_res = op_args.shard->db_slice().FindMutable(op_args.db_cntx, key, OBJ_JSON);
+    if (!it_res.ok()) {
+      return it_res.status();
     }
-  };
 
-  OpStatus status = UpdateEntry(op_args, key, path, cb);
-  if (status != OpStatus::OK) {
-    return status;
-  }
+    PrimeValue& pv = it_res->it->second;
+
+    op_args.shard->search_indices()->RemoveDoc(key, op_args.db_cntx, pv);
 
+    const json::Path& expr = std::get<json::Path>(expression);
+    auto cb = [&vec](optional<string_view>, JsonType* val) {
+      if (val->is_bool()) {
+        bool next_val = val->as_bool() ^ true;
+        *val = next_val;
+        vec.emplace_back(next_val);
+      } else {
+        vec.emplace_back(nullopt);
+      }
+      return false;
+    };
+    json::MutatePath(expr, std::move(cb), pv.GetJson());
+    it_res->post_updater.Run();
+    op_args.shard->search_indices()->AddDoc(key, op_args.db_cntx, pv);
+  } else {
+    auto cb = [&vec](const auto&, JsonType& val) {
+      if (val.is_bool()) {
+        bool current_val = val.as_bool() ^ true;
+        val = current_val;
+        vec.emplace_back(current_val);
+      } else {
+        vec.emplace_back(nullopt);
+      }
+    };
```

Comment:
cb are the same but optional type, I think you can implement one callback via another or do a pointer dereference with constexpr if

Reply:
it's not worth it since it's a transitional code and once we are confident it's robust we will delete the old one

Reply:
any case I have approved it